---
title: "UK Data Processing"
output: html_notebook
editor_options: 
  chunk_output_type: console
---

```{r}
path_to_data <- "~/Documents/PhD_Project/Data/"
```

```{r}
### Reading in the Accession Codes, Population and the Sequence Clusters
UK_metadata <- read.csv("~/Documents/PhD_Project/Data/StrepPneumo_UK/metadata.csv")




UK_ids <- UK_metadata$id
UK_ids_dict <- 1:length(UK_ids)
names(UK_ids_dict) <- UK_ids

UK_id_acc_dict <- UK_ids
names(UK_id_acc_dict) <- UK_metadata$Accession

#Isolates <- Samples_accCodes$`Isolate Name`
#Isolates_dict <- 1:length(Isolates)
#names(Isolates_dict) <- Isolates

#Isolate_from_Mass_dict <- rep(0, length(Isolates))
#names(Isolate_from_Mass_dict) <- Isolates
#Isolate_from_Mass_dict[Mass_Isolates] <- 1

split_winter <- function(winter){
  strsplit(winter,"\\/")[[1]][1]
}

UK_SeqYear_dict <- sapply(UK_metadata$Winter, split_winter)
#names(UK_SeqYear_dict) <- UK_ids
names(UK_SeqYear_dict) <- UK_metadata$Accession

UK_meta_IDtoInd <- 1:nrow(UK_metadata)
names(UK_meta_IDtoInd) <- UK_metadata$Accession

UK_time_points <- sort(unique(UK_SeqYear_dict))

UK_winters <- sort(unique(UK_metadata$Winter))
```

```{r}
saveRDS(UK_metadata, file = "UK_metadata.rds")
```

# use UKjoined data because that should include all sequences that I have metadata for (but there are many, many more that I cannot use...)
```{r}
PP_all <- read.csv(paste(path_to_data, "StrepPneumoUKjoined/PopPUNK/poppunk_clusters_clusters.csv",sep = ""))
no_PP_all <- length(unique(PP_all$Cluster)) 
```

```{r}
PP_taxon_ind_dict <- 1:nrow(PP_all)
names(PP_taxon_ind_dict) <- PP_all$Taxon

PP_is_UK_dict <- rep(FALSE,nrow(PP_all))
names(PP_is_UK_dict) <- PP_all$Taxon

for (i in 1:length(UK_metadata$Accession)) {
  PP_is_UK_dict[UK_metadata$Accession[i]] <- TRUE
}
# seq ERR657323 is only in metadata, not in PP info. Need to remove that one
UK_PP <- PP_all[PP_is_UK_dict,]
UK_PP <- UK_PP[-nrow(UK_PP),]
no_UK_PP <- length(unique(UK_PP$Cluster)) 
# 59 clusters

UK_seq_clusters_dict <- UK_PP$Cluster
names(UK_seq_clusters_dict) <- UK_PP$Taxon
```

```{r}
# Add Serotype Information
UK_Serotype_dict <- UK_metadata$Serotype
names(UK_Serotype_dict) <- UK_metadata$Accession

UK_PP$Serotype <- UK_Serotype_dict[UK_PP$Taxon]

UK_PP$Time <- UK_SeqYear_dict[UK_PP$Taxon]

#UK_VT_dict <- UK_metadata$Vaccine
#names(UK_VT_dict) <- UK_metadata$Accession

#PCV7_dict <- c("NVT", "VT", "NVT")
#names(PCV7_dict) <- c("NVT", "PCV7", "PCV13")

#PCV13_dict <- c("NVT", "VT", "VT")
#names(PCV13_dict) <- c("NVT", "PCV7", "PCV13")

#UK_PP$VT_PCV7 <- PCV7_dict[UK_VT_dict[UK_PP$Taxon]]
#UK_PP$VT_PCV13 <- PCV13_dict[UK_VT_dict[UK_PP$Taxon]]

# create dictionary for VTs of the PCV10
# 1, 4, 5, 6b, 7f, 9V, 14, 18c, 19f and 23f serotypes (source: NCBI)
#PCV7: 4, 6B, 9V, 14, 18C, 19F and 23F. (source: https://journals.plos.org/plospathogens/article?id=10.1371/journal.ppat.1006966#sec018)
# PCV13: (1, 3, 4, 5, 6A and 6B, 7F, 9V, 14, 19A, and 19F, 18C, and 23F) (source: https://www.ncbi.nlm.nih.gov/books/NBK507794/)
PCV13_VTs <- rep("NVT",length(unique(UK_metadata$Serotype)))
names(PCV13_VTs) <- unique(UK_metadata$Serotype)
PCV13_VTs[c("1", "3", "4","5","6A","6B", "7F", "9V", "14", "18C", "19A", "19F", "23F")] <- "VT"

# add 6A to PCV7 because there is strong cross-immunity btw PVC7 and 6A (4.Croucher, N. J. et al. Population genomics of post-vaccine changes in pneumococcal epidemiology. Nat. Genet. 45, 656–663 (2013).)
PCV7_VTs <- rep("NVT",length(unique(UK_metadata$Serotype)))
names(PCV7_VTs) <- unique(UK_metadata$Serotype)
PCV7_VTs[c("4","6A", "6B", "9V", "14", "18C", "19F", "23F")] <- "VT"

UK_PP$VT_PCV7 <- PCV7_VTs[UK_PP$Serotype]
UK_PP$VT_PCV13 <- PCV13_VTs[UK_PP$Serotype]

saveRDS(UK_PP,"UK_PP.rds")
```

# save sequencing year information
```{r}
UKPP_time_points <- unique(sort(UK_PP$Time))
# [1] "2006" "2007" "2008" "2009" "2010" "2011" "2012"

seq_no_per_year <- rep(0, length(UK_time_points))
names(seq_no_per_year) <- UK_winters
for (i in 1:length(UK_time_points)) {
  seq_no_per_year[i] <- length(which(UK_metadata$Winter==UK_winters[i]))
}
barplot(seq_no_per_year, xlab = "Winters", ylim = c(0, 250), ylab = "Sequence Counts", col = rainbow(7))
# introduction of PCV7 in 2006 and PCV13 in 2010
# number of sequences is pretty constant

```

# read in ggCaller results
```{r}
UK_ggCaller <- read.csv(paste(path_to_data, "StrepPneumo_UK3/ggCaller_output/gene_presence_absence.csv", sep = ""), header=FALSE)
```

```{r}
# converting the gene presence absence matrix into a boolean df (0 = gene not present, 1 = gene present)
convert_to_bool <- function(x){
  if (x=="") 0 else 1
}

UK_ggCaller_bool <- UK_ggCaller[,c(-2,-3)]
UK_ggCaller_bool[-1,-1] <- apply(UK_ggCaller_bool[-1,-1],c(1,2), convert_to_bool)

#ggC_matches <- regmatches(UK_ggCaller_bool[1,-1],regexpr("^GCF\\_[0-9]+\\.1",UK_ggCaller_bool[1,-1]))
#ggC_matches_tr <- assembly_dict[ggC_matches]

#gsub_apply <- function(x){
#    ggc_match <- regmatches(x,regexpr("^GCF\\_[0-9]+\\.1",x))
#    ggc_match2 <- gsub("\\.","_",ggc_match)
#    gsub(paste(".*",ggc_match,".*",sep = ""), assembly_dict[ggc_match2],x)
#}

#UK_ggCaller_bool[1,-1] <- sapply(UK_ggCaller_bool[1,-1], gsub_apply)
colnames(UK_ggCaller_bool) <- UK_ggCaller_bool[1,]

# filter ggCaller results by the ones I have metadata for
UK_ggCallerPP_bool <- data.frame(matrix(0, nrow = nrow(UK_ggCaller_bool), ncol = length(UK_PP$Taxon)+1))
for (i in 1:length(UK_PP$Taxon)) {
  UK_ggCallerPP_bool[-1,(i+1)] <- UK_ggCaller_bool[-1,UK_PP$Taxon[i]]
}


UK_ggCallerPP_bool[-1,-1] <- UK_ggCaller_bool[-1,c(UK_PP$Taxon)]
colnames(UK_ggCallerPP_bool) <- c("Gene",UK_PP$Taxon)
UK_ggCallerPP_bool[1,] <- c("Gene",UK_PP$Taxon)
UK_ggCallerPP_bool[,1] <- UK_ggCaller_bool[,1]
```

```{r}
saveRDS(UK_ggCallerPP_bool, "UK_ggCaller_bool.rds")
```

# split ggCaller gene presence absence matrix into different time points
```{r}
UK_ggCaller_byYear <- vector(mode = "list", length = length(UK_time_points))

for (i in 1:length(UK_time_points)) {
  UK_ggCaller_year <- data.frame(matrix(0, nrow = nrow(UK_ggCallerPP_bool), ncol = length(which(UK_metadata$Winter==UK_winters[i]))+1)) # initialize data frame
  UK_ggCaller_year[1,-1] <- UK_ggCallerPP_bool[1,c(FALSE,UK_SeqYear_dict[unlist(UK_ggCallerPP_bool[1,-1])]==UK_time_points[i])] # fill in first row (seq names)
  UK_ggCaller_year[-1,1] <- UK_ggCallerPP_bool[-1,1] # fill in first column (gene cluster names)
  UK_ggCaller_year[-1,-1] <- UK_ggCallerPP_bool[-1,c(FALSE,UK_SeqYear_dict[unlist(UK_ggCallerPP_bool[1,-1])]==UK_time_points[i])] # fill in presence absence information
  #print(i)
  UK_ggCaller_byYear[[i]] <- UK_ggCaller_year
}
```

# find intermediate frequency genes
```{r}
sum_as_int <- function(x){
  sum(as.integer(x))
}
UK_time_points_preVac <- 1
UK_gene_freq_preVac <- rep(0, nrow(UK_ggCaller_byYear[[1]])-1)
preVac_datapoints <- 0
for (i in 1:length(UK_time_points_preVac)) {
  UK_gene_freq_preVac <- UK_gene_freq_preVac + apply(UK_ggCaller_byYear[[i]][-1,-1],1, sum_as_int)
  preVac_datapoints <- preVac_datapoints + length(UK_ggCaller_byYear[[i]][1,-1])
}
UK_gene_freq_preVac <- UK_gene_freq_preVac / preVac_datapoints

UK_gene_filter <- as.integer(UK_gene_freq_preVac<=0.95 & UK_gene_freq_preVac>=0.05)
# sum(UK_gene_filter)
# 1614 intermediate-frequency genes (of 6526)
```

# create intermediate frequency dataframes
```{r}
UK_ggCaller_byYear_intermed <- vector(mode = "list", length = length(UKPP_time_points))

for (i in 1:length(UKPP_time_points)) {
  UK_ggCaller_year <- data.frame(matrix(0, nrow = sum(UK_gene_filter)+1,ncol = length(which(UK_metadata$Year==UKPP_time_points[i]))+1))
  if(ncol(UK_ggCaller_byYear[[i]])>1){
    UK_ggCaller_year <-  UK_ggCaller_byYear[[i]][c(1,which(UK_gene_filter==1)+1),]
  }
  else{
    UK_ggCaller_year <-  data.frame(matrix(UK_ggCaller_byYear[[6]][c(1,which(UK_gene_filter==1)+1),1], nrow = sum(UK_gene_filter)+1,ncol = 1))
  }
  UK_ggCaller_byYear_intermed[[i]] <- UK_ggCaller_year 
}

UK_ggCaller_intermed <- data.frame(matrix(0, nrow = sum(UK_gene_filter)+1,ncol = ncol(UK_ggCallerPP_bool)))
UK_ggCaller_intermed <-  UK_ggCallerPP_bool[c(1,which(UK_gene_filter==1)+1),]

# create pre_Vac intermed frequency data frame
UK_ggCaller_intermed_preVac <- data.frame(matrix(0, nrow = sum(UK_gene_filter)+1,ncol = (sum(unlist(lapply(UK_ggCaller_byYear_intermed,ncol))[1:length(UK_time_points_preVac)]) - length(UK_time_points_preVac) + 1)))
col_count <- 2
for (i in 1:length(UK_time_points_preVac)) {
  if(ncol(UK_ggCaller_byYear_intermed[[i]])>1){
    UK_ggCaller_intermed_preVac[,(col_count:(col_count + ncol(UK_ggCaller_byYear_intermed[[i]])-1))] <- UK_ggCaller_byYear_intermed[[i]][,-1]
    col_count <- col_count + ncol(UK_ggCaller_byYear_intermed[[i]])
  }
}
UK_ggCaller_intermed_preVac[,1] <-  UK_ggCaller_byYear_intermed[[1]][,1]
```

# create consensus gene presence absence matrices for the clusters
```{r}
cons_genomes <- function(x){
  as.double(ceiling(median(as.integer(x))))
}

# overall
UK_ggCaller_intermed_consensus <- data.frame(matrix(0, nrow = sum(UK_gene_filter)+1, ncol = no_UK_PP+1))
UK_ggCaller_intermed_consensus[1,-1] <- paste("SeqCl_",unique(UK_PP$Cluster),sep = "")

for (i in unique(UK_PP$Cluster)) {
  UK_ggCaller_intermed_consensus[-1,(which(unique(UK_PP$Cluster) == i))+1] <- apply(as.matrix(UK_ggCaller_intermed[-1,c(FALSE,UK_seq_clusters_dict[unlist(UK_ggCaller_intermed[1,-1])]==i)]), 1, cons_genomes)

}
UK_ggCaller_intermed_consensus[,1] <- UK_ggCaller_intermed[,1]

#by year
UK_ggCaller_byYear_intermed_consensus <- vector(mode = "list", length = length(UK_time_points))

for (i in 1:length(UK_time_points)) {
  UK_ggCaller_year <- data.frame(matrix(0, nrow = sum(UK_gene_filter)+1,ncol = no_UK_PP+1))
  UK_ggCaller_year[1,-1] <- paste("SeqCl_",unique(UK_PP$Cluster),sep = "")
  for (j in unique(UK_PP$Cluster)) {
    UK_ggCaller_year[-1,(which(unique(UK_PP$Cluster) == j))+1] <- apply(as.matrix(UK_ggCaller_byYear_intermed[[i]][-1,c(FALSE,UK_seq_clusters_dict[unlist(UK_ggCaller_byYear_intermed[[i]][1,-1])]==j)]), 1, cons_genomes)
  }
  UK_ggCaller_year[,1] <- UK_ggCaller_byYear_intermed[[i]][,1]
  UK_ggCaller_byYear_intermed_consensus[[i]] <- UK_ggCaller_year
}

saveRDS(UK_ggCaller_intermed_consensus,"UK_ggCaller_intermed_consensus.rds")
```

# Calculate VT/NVT for clusters
```{r}
# introduction of PCV7 in 2006 and PCV13 in 2010

# create dictionary for VTs of the PCV7
# 1, 4, 5, 6b, 7f, 9V, 14, 18c, 19f and 23f serotypes (source: NCBI)
#PCV7: 4, 6B, 9V, 14, 18C, 19F and 23F. (source: https://journals.plos.org/plospathogens/article?id=10.1371/journal.ppat.1006966#sec018)
# PCV13: (1, 3, 4, 5, 6A and 6B, 7F, 9V, 14, 19A, and 19F, 18C, and 23F) (source: https://www.ncbi.nlm.nih.gov/books/NBK507794/)
PCV13_VTs <- rep(0,length(unique(UK_metadata$Serotype)))
names(PCV13_VTs) <- unique(UK_metadata$Serotype)
PCV13_VTs[c("1", "3", "4","5","6A","6B", "7F", "9V", "14", "18C", "19A", "19F", "23F")] <- 1

PCV7_VTs_no6A <- rep(0,length(unique(UK_metadata$Serotype)))
names(PCV7_VTs_no6A) <- unique(UK_metadata$Serotype)
PCV7_VTs_no6A[c("4","6B", "9V", "14", "18C", "19F", "23F")] <- 1

# add 6A to PCV7 because there is strong cross-immunity btw PVC7 and 6A (4.Croucher, N. J. et al. Population genomics of post-vaccine changes in pneumococcal epidemiology. Nat. Genet. 45, 656–663 (2013).)
PCV7_VTs <- rep(0,length(unique(UK_metadata$Serotype)))
names(PCV7_VTs) <- unique(UK_metadata$Serotype)
PCV7_VTs[c("4","6A", "6B", "9V", "14", "18C", "19F", "23F")] <- 1

UK_SeqToSero_dict <- UK_metadata$Serotype
names(UK_SeqToSero_dict) <- UK_metadata$Accession

# calculate Vaccine Type consensus for clusters
UK_VT_no6A <- rep(0, no_UK_PP)

for (i in unique(UK_PP$Cluster)){
  UK_VT_no6A[(which(unique(UK_PP$Cluster) == i))] <- ceiling(median(as.integer(PCV7_VTs_no6A[UK_SeqToSero_dict[UK_PP[UK_PP$Cluster == i,"Taxon"]]])))
}

# calculate Vaccine Type consensus for clusters
UK_VT <- rep(0, no_UK_PP)

for (i in unique(UK_PP$Cluster)){
  UK_VT[(which(unique(UK_PP$Cluster) == i))] <- ceiling(median(as.integer(PCV7_VTs[UK_SeqToSero_dict[UK_PP[UK_PP$Cluster == i,"Taxon"]]])))
}
# 1 means VT
# 0 means NVT
# if cluster is 50/50 it will count as VT

# calculate Vaccine Type consensus for clusters for PCV13
UK_VT2 <- rep(0, no_UK_PP)

for (i in unique(UK_PP$Cluster)){
  UK_VT2[(which(unique(UK_PP$Cluster) == i))] <- ceiling(median(as.integer(PCV13_VTs[UK_SeqToSero_dict[UK_PP[UK_PP$Cluster == i,"Taxon"]]])))
}
```

```{r}
#save VTs
saveRDS(UK_VT_no6A, file = "UK_VT_no6A.rds")
saveRDS(UK_VT, file = "UK_VT.rds")
saveRDS(UK_VT2, file = "UK_VT2.rds")
```

# calculate cluster frequencies for different years
```{r}
UK_cluster_freqs <- vector(mode = "list", length = length(UK_time_points))

for(j in 1:length(UK_time_points)){
  UK_cluster_freqs[[j]] <- rep(NA, no_UK_PP)
  for (i in unique(UK_PP$Cluster)){
    UK_cluster_freqs[[j]][which(unique(UK_PP$Cluster) == i)] <- length(which(UK_metadata[UK_meta_IDtoInd[UK_PP[UK_PP$Cluster == i,"Taxon"]],]$Year==UK_time_points[j]))
  }
}

for(j in 1:length(UK_time_points)){
  if(sum(UK_cluster_freqs[[j]])==0){
    UK_cluster_freqs[[j]] <- rep(NA, no_UK_PP)
  }
}

#UK_cluster_freqs[[6]] <- rep(NA, no_UK_PP)
# 901 data points in total
UK_cluster_freqs_preVac <- rep(0, no_UK_PP)
for (i in 1:length(UK_time_points_preVac)) {
  UK_cluster_freqs_preVac <- UK_cluster_freqs_preVac + replace(UK_cluster_freqs[[i]],which(is.na(UK_cluster_freqs[[i]])),0)
}

for (i in 1:length(UK_time_points)) {
  file_name <- paste("UK_cluster_freqs_", as.character(i),".rds", sep = "")
  saveRDS(UK_cluster_freqs[[i]], file_name)
}


# do the same for Strain x Serotype model
UK_cluster_freqs_sero <- vector(mode = "list", length = length(UK_time_points))

for(j in 1:length(UK_time_points)){
  UK_cluster_freqs_sero[[j]] <- matrix(NA, ncol =  length(unique(UK_PP$Serotype)), nrow = no_UK_PP)
  for (i in unique(UK_PP$Cluster)){
    for (k in unique(UK_PP$Serotype)) {
      UK_cluster_freqs_sero[[j]][which(unique(UK_PP$Cluster) == i),which(unique(UK_PP$Serotype) == k)] <- length(which(UK_metadata[UK_meta_IDtoInd[UK_PP[UK_PP$Cluster == i,"Taxon"]],][UK_metadata[UK_meta_IDtoInd[UK_PP[UK_PP$Cluster == i,"Taxon"]],]$Serotype==k,]$Year==UK_time_points[j]))
    }
  }
}
for(j in 1:length(UK_time_points)){
  if(sum(UK_cluster_freqs_sero[[j]])==0){
    UK_cluster_freqs_sero[[j]] <- matrix(NA, ncol =  length(unique(UK_PP$Serotype)), nrow = no_UK_PP)
  }
}

for (i in 1:length(UK_time_points)) {
  file_name <- paste("UK_cluster_freqs_sero_", as.character(i),".rds", sep = "")
  saveRDS(UK_cluster_freqs_sero[[i]], file_name)
}


```

#create initial population
```{r}
### create initial population that is based on the first time point data set
# sample from it with an Poisson process
expand_factor <- 15000 / sum(UK_cluster_freqs_preVac)
exp_noise <- 10
UK_model_start_pop <- (sapply((UK_cluster_freqs_preVac + rexp(n = length(UK_cluster_freqs_preVac), rate = exp_noise)) * expand_factor, rpois, n=1))

# visual check for the "sampling"
plot(UK_cluster_freqs_preVac/sum(UK_cluster_freqs_preVac))
points(UK_model_start_pop/sum(UK_model_start_pop), col = "red")

expand_factor <- 15000 / sum(UK_cluster_freqs[[1]])
exp_noise <- 10
UK_model_start_pop <- (sapply((UK_cluster_freqs[[1]] + rexp(n = length(UK_cluster_freqs[[1]]), rate = exp_noise)) * expand_factor, rpois, n=1))

# visual check for the "sampling"
plot(UK_cluster_freqs[[1]]/sum(UK_cluster_freqs[[1]]))
points(UK_model_start_pop/sum(UK_model_start_pop), col = "red")

saveRDS(UK_model_start_pop, "UK_model_start_pop.rds")
```


# calculate delta ranking
# I will do last data point - all pre_vac data points combined
```{r}
# For strong and weak selection in the model, I need to calculate the beta statistics.

# calculate gene frequencies first, separate for three time points
UK_gene_freq_preVac <- apply(UK_ggCaller_intermed_preVac[-1,-1], 1, sum_as_int)
UK_gene_freq_postVac <- apply(UK_ggCaller_byYear_intermed[[length(UK_ggCaller_byYear_intermed)]][-1,-1], 1, sum_as_int) + apply(UK_ggCaller_byYear_intermed[[length(UK_ggCaller_byYear_intermed)-1]][-1,-1], 1, sum_as_int) + apply(UK_ggCaller_byYear_intermed[[length(UK_ggCaller_byYear_intermed)-2]][-1,-1], 1, sum_as_int)


# first, calculate pre/peri and post vacc frequencies of genes:
UK_gene_freq_preVac_rel <- UK_gene_freq_preVac/sum(UK_gene_freq_preVac)
UK_gene_freq_postVac_rel <- UK_gene_freq_postVac/sum(UK_gene_freq_postVac)

# calculate delta statistic (refer to Corander et al. for more info)
UK_delta_data <- (UK_gene_freq_postVac_rel - UK_gene_freq_preVac_rel) ^ 2 / (1 - UK_gene_freq_preVac_rel * (1 - UK_gene_freq_preVac_rel))
UK_delta_ranking <- rank(UK_delta_data)

saveRDS(UK_delta_ranking,"UK_delta_ranking.rds")
```

### Data processing for PPxSero model
```{r}
PopPUNK_clusters <- readRDS("UK_PP.rds")
UK_model_start_pop <- readRDS("UK_model_start_pop.rds")
mass_clusters <- length(unique(PopPUNK_clusters$Cluster))
UK_avg_cluster_freq <- rep(1/mass_clusters, mass_clusters)


PPxSero_matrix <- matrix(0, nrow = length(unique(PopPUNK_clusters$Cluster)), ncol = length(unique(PopPUNK_clusters$Serotype)))
# sort(unique(PopPUNK_clusters$Serotype))
# [1] "10A"   "11A"   "14"    "15A"   "15B/C" "15F"   "16F"   "17F"   "18C"   "19A"   "19F"   "21"    "22F"   "23A"
#[15] "23B"   "23F"   "3"     "31"    "33F"   "34"    "35B"   "35F"   "37"    "38"    "6A"    "6B"    "6C"    "7C" 
#[29] "7F"    "9N"    "9V"    "NT"

SerotypeToIndex_dict <- 1:length(unique(PopPUNK_clusters$Serotype))
names(SerotypeToIndex_dict) <- sort(unique(PopPUNK_clusters$Serotype))

IndexToSerotype_dict <- sort(unique(PopPUNK_clusters$Serotype))
names(IndexToSerotype_dict) <- 1:length(unique(PopPUNK_clusters$Serotype))

PPclustToIndex_dict <- 1:length(unique(PopPUNK_clusters$Cluster))
names(PPclustToIndex_dict) <- sort(unique(PopPUNK_clusters$Cluster))

for (i in 1:nrow(PopPUNK_clusters)) {
  PPxSero_matrix[PPclustToIndex_dict[as.character(PopPUNK_clusters$Cluster[i])],SerotypeToIndex_dict[PopPUNK_clusters$Serotype[i]]] <- 1
}
# matrix shows information whether PP and serotype co-occur in the dataset (1=TRUE)
# this can be used as the migration matrix:
#PPxSero_matrix_prob <- PPxSero_matrix / sum(PPxSero_matrix)
#PPxSero_matrix_prob <- matrix(sapply(PPxSero_matrix_prob,as.double), nrow = nrow(PPxSero_matrix), ncol = ncol(PPxSero_matrix))
#saveRDS(PPxSero_matrix_prob, "Navajo_PPsero_mig.rds")


PPxSero_matrix_prob2 <-  matrix(0, nrow = length(unique(PopPUNK_clusters$Cluster)), ncol = length(unique(PopPUNK_clusters$Serotype)))
for (i in 1:length(unique(PopPUNK_clusters$Cluster))) {
  PPxSero_matrix_prob2[i,] <- ((PPxSero_matrix[i,] * UK_avg_cluster_freq[i])/sum(PPxSero_matrix[i,]))
}
saveRDS(PPxSero_matrix_prob2, "UK_PPsero_mig.rds")
# new migration matrix that is closer to original one

# start population:
PPsero_startpop <- matrix(0, nrow = length(unique(PopPUNK_clusters$Cluster)), ncol = length(unique(PopPUNK_clusters$Serotype)))

for (i in which(PopPUNK_clusters$SeqYear=="1998")) {
  PPsero_startpop[PPclustToIndex_dict[PopPUNK_clusters$Cluster[i]],SerotypeToIndex_dict[PopPUNK_clusters$Serotype[i]]] <- PPsero_startpop[PPclustToIndex_dict[PopPUNK_clusters$Cluster[i]],SerotypeToIndex_dict[PopPUNK_clusters$Serotype[i]]] + 1
}
#PPsero_startpop <- PPsero_startpop * 15000 / sum(PPsero_startpop)


PP_expand_factor <- 15000 / sum(PPsero_startpop)
exp_noise <- 10
#PPsero_startpop_corr <- matrix(sapply((PPsero_startpop + matrix(rexp(n = ncol(PPsero_startpop) * nrow(PPsero_startpop), rate = exp_noise), nrow= nrow(PPsero_startpop), ncol = ncol(PPsero_startpop))) * PP_expand_factor, rpois, n=1), nrow= nrow(PPsero_startpop), ncol = ncol(PPsero_startpop))
#PPsero_startpop_corr <- matrix(sapply(PPsero_startpop_corr,as.double), nrow= nrow(PPsero_startpop), ncol = ncol(PPsero_startpop))
#saveRDS(PPsero_startpop_corr, "PPsero_startpop.rds")

# this lead to a start population that is way too evenly distributed across the different clusters
# instead, I could maybe use the existing PP_model_start_pop and distribute it to the different serotypes?
PPsero_startpop2 <-  matrix(0, nrow = length(unique(PopPUNK_clusters$Cluster)), ncol = length(unique(PopPUNK_clusters$Serotype)))
for (i in 1:length(unique(PopPUNK_clusters$Cluster))) {
  PPsero_startpop2[i,] <- round((PPxSero_matrix[i,] * UK_model_start_pop[i])/sum(PPxSero_matrix[i,]))
}
saveRDS(PPsero_startpop2, "UK_PPsero_startpop.rds")
# this is very close to the original PP_model_start_pop
# maybe I am giving out too much information because the serotype distribution is exactly the real one?

# serotype VT vector
SeroVT <- rep(0, length(unique(PopPUNK_clusters$Serotype)))
for (i in 1:nrow(PopPUNK_clusters)) {
  SeroVT[SerotypeToIndex_dict[PopPUNK_clusters$Serotype[i]]] <- as.integer(PopPUNK_clusters$VT_PCV7[i]=="VT")
}
saveRDS(SeroVT, "UK_SeroVT.rds")

# serotype VT vector PCV13
SeroVT_PCV13 <- rep(0, length(unique(PopPUNK_clusters$Serotype)))
for (i in 1:nrow(PopPUNK_clusters)) {
  SeroVT_PCV13[SerotypeToIndex_dict[PopPUNK_clusters$Serotype[i]]] <- as.integer(PopPUNK_clusters$VT_PCV13[i]=="VT")
}
saveRDS(SeroVT_PCV13, "UK_SeroVT_PCV13.rds")
```
